File: problems/045/045.f90
PROGRAM F045

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    INTEGER (KIND=8), PARAMETER :: MAX_N     = 1E5

    REAL    (KIND=8), PARAMETER :: TOLERANCE = 1E-5

    REAL    (KIND=8)            :: SQRT_TERM,  TN

    INTEGER :: I

    DO I = 286, MAX_N

        TN = 0.5D0 * I * (I + 1.0D0)

        SQRT_TERM = DSQRT(1.D0 + (24.D0 * TN))

        IF ((SQRT_TERM - REAL(FLOOR(SQRT_TERM, 8), 8)) > TOLERANCE) CYCLE

        IF (MOD(FLOOR(SQRT_TERM, 8) + 1.0D0, 6.0D0) > TOLERANCE) CYCLE

        SQRT_TERM = DSQRT(1.D0 + (8.D0 * TN))

        IF ((SQRT_TERM - REAL(FLOOR(SQRT_TERM, 8), 8)) > TOLERANCE) CYCLE

        IF (MOD(FLOOR(SQRT_TERM, 8) + 1.0D0, 4.0D0) <= TOLERANCE) EXIT

    END DO

    PRINT ('(I0)'), FLOOR(TN, 8)

END PROGRAM F045

File: problems/045/045.c
#include <math.h>
#include <stdio.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    long int tn;

    const int MAX_N = 1E5;

    for (int i = 286; i <= MAX_N; i++)
    {
        tn = 0.5 * i * (i + 1);

        double sqrt_term = sqrt(1 + (24 * tn));

        if (sqrt_term != (long int)sqrt_term)
            continue;

        if ((((long int)sqrt_term + 1) % 6) != 0)
            continue;

        sqrt_term = sqrt(1 + (8 * tn));

        if (sqrt_term != (long int)sqrt_term)
            continue;

        if ((((long int)sqrt_term + 1) % 4) == 0)
            break;
    }

    printf("%ld\n", tn);

    return 0;
}
